### ISSUE TITLE
Supply Mule stops working after one Trade Run

### ISSUE THREAD

**Issue by SM-Tiger** [Misunderstood-Wookiee/Mules-and-Warehosues-Extended#12]
Good Evening,

set up Supply Mules to serve a station and only use Player Stations (All Wares)
After some time they finish their first job (deliver from Player Station to Building Site) and then they will just sit there searching for trades :-/

Same goes for Distribution Mule.

Station Mule working fine, going to set those up now.
Thanks for your work

**Comment by bhayden53** [comment]
The original author had completely replaced one of the vanilla move orders, which might end up causing some of the mules not to undock. Wookie was working on that earlier today but I don't think a fix ever went through. 

I'm poking around with distrib mule this evening, I'll take a look.

In the release versions of the mod on steam and nexus there's a bug in station mule regarding certain wares. Basically it will tend not to move your end products to the warehouse. i.e. hull parts, claytronics, etc. That's fixed in public. I assume wookie will put out a release tomorrow.

**Comment by SM-Tiger** [comment]
Supply Mule seems to be seriously broken.
If i tell them to use Player Stations only they will kinda just sit there searching for trades.
If i allow them to trade with other factions they will start supplying the building sites, but once that is done they will again just sit there searching for trades.
If i tell it to only trade a few wares it seems to be working, but only if allowed to trade with other factions. (but thats exactly what i dont want)

Station Mule is working fine, but while my Wharf wants 800 k ECells they will not go and buy from my Source Station. (Probably because thats only a secondary trading item for them, and because the Source Station doesnt have anything else to trade for that Wharf they will instead not Trade at all)

Set up a Distribution Mule for ECells, Food and Meds now, first trade run looking good, going to see if it keeps going or goes into slumber again.


Really happy with Station Mule, working just as i want it to. They will sometimes Search for Trades, but after some research thats really because the Source Station doesnt have Supplys ready yet or the Target Stations storage is as close as full.


If the supply mule would just work as a dedicated Building Site Supply Mule that would actually be much more appreciated. (Because if i tell them to supply anything they will work on those Building Sites, but once thats done they will start buying expensive Goods for my Wharf, and i am not sure thats what i want at this time. //Money Wise)

**Comment by Misunderstood-Wookiee** [comment]
Correct in one the key locations -  Here is a snippet of the section.
```
<!-- Order is ready to be started -->
			  <!-- In case the ship is docked, first create an immediate undock order
					NB: ships that are docked do not have to park to trade.
					ships that are parked do have to dock to have anything built. --
				<!-- NB: remove releasesignal params in Undock calls if we remove the undock-specific requirement to tell a ship to proceed. -->
				<do_if value="$object.order.trade or $object.order.build or ($object.order.id == 'StationMule') or ($object.order.id == 'TravelMule') or ($object.order.id == 'DistriMule') or ($object.order.id == 'DockAt') or ($object.order.id == 'DockAndWait') or ((($object.order.id == 'Wait') or ($object.order.id == 'WaitForSignal')) and $object.order.$allowdocked and not 
```

Straight out of the orders.base XML file  the same file which made all ships not able to undock during the beta for update 3.0 as  prev stated I know why it screwed up in beta as the script is using 

> <diff> <replace> entire script </replace> </diff> 

Replacing the whole script and thus losing a good 30 + lines of updated code in the newer files whole sections are just missing the new file the script does not work the same way as this old one.

**Comment by bhayden53** [comment]
I haven't looked very closely at supply mule yet, but all of my testing indicates that station, distribution, and travel mules are working as intended. There are some gotchas for travel and distribution mule though.

Most of the time for distribution mule, if it seems like it's not doing anything but it should, it's due to the ratio sliders in the UI. 
Min target ratio: if the ware on the target distribution center is above this no trade will happen. i.e., make the value higher and stuff might start moving.
static ratio: if the ware on the source station is below this no trade will happen. i.e. make the value lower and stuff might start moving.
The default values for distribution mule seem to be best for when you have one major warehouse pushing out wares to 5+ other ones.
Haven't noticed any issues with the max ratio slider yet because I haven't yet been able to full up a bunch of warehouses.

Travel Mule takes the warehouses buy order into account to estimate a profit and attempts to fill the cargo more than 75%. So while it seems like there should be trades, my poking around with stations in the area seemed to indicate that the warehouse had high sell prices which was stopping the mule from finding a "profitable" trade. I'm thinking perhaps a slider parameter or even a total rethink may be in order for what is a valid trade for the travel mule. 

I hope to get to supply mule soon, sorry, this is a lot of work.

**Comment by SM-Tiger** [comment]
Travel Mule doesnt sound so dump to me.
I was getting really furious when i saw trades of the vanilla trader running around with 40 silicon wafers.
Its much more "logistical" to make sure we have at least 75% Cargo Space used.
Maybe set a own minimum price treshold like 110% of base price, instead of having to wait for the factory to allocate enough storage for it to reach this price treshold.
Other way our traders would only start working once NPC traders find our deals lucrative too, stealing us those extra credits. Profitsssssss


On the Supply Mule.
It does seem to be working fine if i am telling him to trade for certain wares and make it only Supply Source. i.E. i set up 10 M and 5 L Traders for my Wharf now buying needed goods from NPC Stations. The L's will only buy Hullparts.
But i had to set up a Warelist manually and make it Supply Source only, just click OK it will take all wares from the universe and instead of working only supplies own building sites. (in a certain distance, it wont supply my Build Defence Station Missions Building Site that are a few Sectors away.)

I feel ya mate. Playing around with other ppls scripts sorting errors out and so on isnt always the easiest way. Sometimes its much easier to just start from scratch.
Have been modding in Rebirth working out a way to turn the magical resource sink into a real economy / war like situation. Just like its in X4 Vanilla now :-D

**Comment by bhayden53** [comment]
That's good info on the supply mule, thanks. Another thing worth mentioning is that the supply mule was written by a totally different author. Wookie combined it in with the original ones for convenience. 

One thing I'm wondering.... have you seen any AI ships trading with the warehouses? I'm not sure yet if they're trading.

Another thing to watch for: if you open the "plan build" menu, a lot of times you can see that every build order for materials is already filled by a planned trade with an AI ship. In that case your suppliers wouldn't have anything to do.

**Comment by SM-Tiger** [comment]
Dont have any Warehouses, so cant tell you anything about that.
Rather Sell the Ship directly, as that will raise the final sell price for the intermediate goods to about 300% :-D

Yes i know about the planned trades in Plan Build Menu.

It seems that Supply Mules that are set to Supply Source Only will ignore Building Sites once they have done their initial Trade Run.
If they are not set to Supply Source Only they will not Supply the Source though :-/

Warelist actually isnt a matter at all.

**Comment by bhayden53** [comment]
ok thanks for narrowing it down, I'll try to look at the supply mule this evening after work

**Comment by macrixen** [comment]
I do not know if this is related or even helps but here is an error code i find in my debug log when the supply mule stops working for me.
`[=ERROR=] 380553.52 Error in AI script supplymule on entity 0x366a4: Division by zero! 0f / 0f
* Expression: ($needs.{$vy}.desiredamount)f / ($needs.{$vy}.owner.cargo.{$needs.{$vy}.ware}.target)f * 1000
* Action: <append_to_list>, line 366
[General] 380553.52 ======================================
[General] 380553.52 ======================================
[=ERROR=] 380553.52 Error in AI script supplymule on entity 0x366a4: Unreliable comparison of non-numeric script values! 0 < [ware.claytronics,ware.energycells,ware.foodrations,ware.hullparts,ware.medicalsupplies]
* Expression: $Supplier.count gt 0 and $wareBasket gt 0
* Action: <do_if>, line 447
[General] 380553.52 ======================================`

also it seems to supply the build as long as it is the first time it builds the station but any changes or additions it seems to hang.

**Comment by bhayden53** [comment]
I suspect $wareBasket gt 0 should be $wareBasket.count gt 0.

not sure if the divide by zero is breaking anything.

I could see both of these being major issues, not sure yet.

**Comment by bhayden53** [comment]
it also looks like supply mule is the only one using the custom quicksort script

**Comment by SM-Tiger** [comment]
yes the line 447 is easy fix.

for line 366 i dont really get whats the 'f' doing behind the ( )
($needs.{$vy}.desiredamount)f / ($needs.{$vy}.owner.cargo.{$needs.{$vy}.ware}.target)f * 1000
this is pretty much the same as for StationMule trying to find out how many % if allocated storage is empty. my guess would be, that the 'f' is breaking the math.
And its probably breaking for the Mule to know which goods to search for first.

This is how station mule does priorizing what to serve first:
That way you would also get rid of the quicksort
`
```

> 						<do_all exact="$needs.count" counter="$nSort">
> 							<set_value name="$someTrade" exact="$needs.{$nSort}" />
> 							<set_value name="$isMadeAndUsed" exact="$someTrade.owner.products.{$someTrade.ware}.exists" />
> 							<set_value name="$isTradeWare" exact="$someTrade.owner.tradewares.{$someTrade.ware}.exists" />
> 							<set_value name="$cargoRatio" exact="($someTrade.owner.cargo.{$someTrade.ware}.count * 100 / [$someTrade.owner.cargo.{$someTrade.ware}.target, 1].max)" />
> 							<do_if value="(not $isMadeAndUsed) and (not $isTradeWare) and $cargoRatio lt 10">
> 								<append_to_list name="$outerList" exact="$someTrade" />
> 							</do_if>
> 							<do_elseif value="(not $isMadeAndUsed) and (not $isTradeWare) and $cargoRatio lt 25">
> 								<append_to_list name="$prio2" exact="$someTrade" />
> 							</do_elseif>
> 							<do_elseif value="(not $isTradeWare) and $cargoRatio lt 50">
> 								<append_to_list name="$prio3" exact="$someTrade" />
> 							</do_elseif>
> 							<do_elseif value="(not $isTradeWare) and $cargoRatio lt 75">
> 								<append_to_list name="$prio4" exact="$someTrade" />
> 							</do_elseif>
> 							<do_else>
> 								<append_to_list name="$prio5" exact="$someTrade" />
> 							</do_else>
> 						</do_all>

```

**Comment by bhayden53** [comment]
I fixed both of these issues and it doesnt fix the mule stopping after one run, unfortunately. 

The f at the end is an explicit float typecasting. 

Now I have to dig way deeper.

**Comment by bhayden53** [comment]
I suspect the problem is related to some sort of variable persistence across supply runs, I'm just trying to print debug everything and find it

**Comment by bhayden53** [comment]
so I assigned a supply mule to a station and it added ore and methane to it's warelist? Those can only be supplied by miners. And anyway the station isn't even built yet.

I need to go through this line by line, this mule is in a bad place.

**Comment by bhayden53** [comment]
I am suspecting that the warelist is the problem, after one trip the only ware in the warelist that's actually needed in build storage is energy cells, and the build storage is already full for that, so there's nothing for the mule to do. I have a list that's energy cells, graphene, ore, methane, and refined metals. For a station that has no modules completed.

**Comment by bhayden53** [comment]
There should be a fix rolling out for this soon. The logic used to construct the ware list was messy and not working correctly. My fix greatly simplifies it, but the supply mule can no longer create trade orders on a player warehouse, if you ever relied on it for that.

I will try to get the trade order creation back into a future release, but for now I wanted it back to being functional.

main fix is in cf6aa10a8a25239e844147119ff887b130cc4c19 with minor bug fixes in later commits

**Comment by macrixen** [comment]
I started getting this when my supply mule suddenly got stuck on the last delivery. Looking in the orders it does not show the Dock command but has the Execute Trade and ends up spamming fund transfer.
```
[=ERROR=] 404283.60 Error in AI script supplymule on entity 0x8bae57: Property lookup failed: $needs
* Expression: $needs.count == 0
* Action: <do_if>, line 269
```

**Comment by Misunderstood-Wookiee** [comment]
Re-Open until further debug is done.

**Comment by bhayden53** [comment]
Once the ship has arrived there I think the dock command goes away? Anyway none of that is handled at all in this script. All we do is make the trade order, vanilla scripts handle how it is then actually executed.

needs is set when the "give builders priority" box is on, and is the result of a search for buy orders on the station that is assigned. Can you check what settings you had on the mule? I've only tested with that box turned on. The needs error should be solved by having it checked, for now.

I really have no idea what would cause the behavior you describe.

**Comment by bhayden53** [comment]
the only thing I can think of is that somehow between the time we got the trade and the time the ship got there, the trade was no longer valid. If you can, please provide more context for what the ship was doing.

**Comment by macrixen** [comment]
the dock command was never added. I cleared and gave the mule back it's setup with the wares added the execute trade but somehow does not add the dock and dock and wait commands. I should also note i did not have enough credits to complete the full trade. Whether that has any effect to it or not...*shrug*. The ship was flying next to the build site, not anywhere near where the ship would be going to buy the wares. Will see if I can re-create the problem.

**Comment by bhayden53** [comment]
I found the bug with the needs object not existing. It was not introduced by me, it's been in the code since we got it. No idea whether it solves your problem.

I'm also introducing explicit checking that the trade is available just before making the trade order. It's unlikely to be the problem but good practice since there is always the potential for a race condition.

If you're launching in debug, can you also generate the script log files i.e. with -scriptlogfiles ? It will have a name starting with supplymule

**Comment by bhayden53** [comment]
addressed the needs issue here 9a2982d7fd15ed29c6a0739d0e176f706164d047

Do continue to report any funny business with supply mule.

**Comment by leoric** [comment]
I somehow got similar bug on 5.1.7 version, everything was fine for more than 24h of playtime, then all supply mules just stopped working. If I load previous save (just 42 minutes difference), everything is still working. Same goes for removing the command and setting it back - everything is fine in old save, not working at all in new save.
Uprading to 5.1.8 didn't solve the issue.

**Comment by bhayden53** [comment]
Supply mule will soon be replaced by a new script that has been written from scratch. So unfortunately, I don't want to sink any of my time into debugging the old script.

What I will say is that I added a fairly long wait at the end of the script, 3 to 5 minutes. If the mule finds nothing to do, it will wait to check again. So they may appear to sit in searching for trades for a long time. 

The first thing to do is to look really carefully into whether or not trades actually exist that the mule should be fulfilling. 95% of the time that I've thought something was wrong with the mules, when I looked into the debug logs, it was working as intended, there was just nothing to do.
